NSF funding pilot phase 2

daisieh opened the pull request:
Addresses https://trello.com/c/ebGo5uBp/478-capture-funding-data-at-submission-phase-2.

daisieh committed:
new class for funder concepts

daisieh committed:
Moving things around between organizations, funders, journals

daisieh committed:
refactoring out redundant arguments in cart methods

daisieh committed:
add organization-updates script to clean up the metadatafieldregistry and conceptmetadatavalue tables

daisieh committed:
add commands to update the organization view

daisieh committed:
optimize imports

daisieh committed:
make DryadJournalConcept a subclass of a more generic DryadOrganizationConcept

daisieh committed:
not sure what happened here

daisieh committed:
change schema to "organization"

daisieh committed:
set journal explicitly, then update total

daisieh committed:
move getSubscriptionPaid

daisieh committed:
setJournal is private

daisieh committed:
if null, set journal info to null

daisieh committed:
change organization view to journal view

daisieh committed:
Add method Concept.searchByMetadata

daisieh committed:
constructor to make a DryadOrganizationConcept from a Concept

daisieh committed:
remove defunct metadata fields

daisieh committed:
search for concepts by metadata

daisieh committed:
get organization concepts by name or customerID

daisieh committed:
find NSF organization concept by customerID

daisieh committed:
clean up tests

daisieh committed:
add sponsor_id to the shopping cart

daisieh committed:
setOrgConcept -> setSponsoringOrg

daisieh committed:
getDryadOrganizationConceptMatchingConceptID

daisieh committed:
pass in context

daisieh committed:
use organization concepts instead of manually looking for concept

daisieh committed:
more funder info

daisieh committed:
update journal/org concepts

daisieh committed:
ImportFunder2Authority

daisieh committed:
change dataone repo url

daisieh committed:
add funding entities to item view

daisieh committed:
update interfaces to use NSF's DryadFunderConcept

daisieh committed:
rearrange package view, add funder info

daisieh committed:
Add funding information to datacite crosswalks

daisieh commented:
In order to set this up on a VM, a bunch of stuff needs to be done. Update your dryad-repo and then execute the following commands:

```
update_schemas.sh
/opt/dryad/bin/dsrun org.dspace.administer.MetadataImporter -f /opt/dryad/config/registries/funder-types.xml
/opt/dryad/bin/dsrun org.dspace.administer.MetadataImporter -f /opt/dryad/config/registries/organization-types.xml
python dspace/etc/upgrades/organization-updates.py
/opt/dryad/bin/dsrun org.datadryad.authority.ImportFunder2Authority
```

The python script updates the database's schemas to reflect the new metadata fields and schemas that were installed in this PR.
The class ImportFunder2Authority imports all of the funder concepts from the FundRef database. If the command gets stuck at "starting", then kill it and restart it: sometimes it can't access the URL. When it works, it will add ~13,300 funder concepts to the database.

daisieh commented:
Once you've set up the VM, you should go check the NSF concept (funder concept, not the journal concept) and add a fake customer ID number for it. You'll have to go to Manage Scheme>Funder>Search and Add Concept to find it.

Then you can test.

• Create a submission without sponsorship by journal but with a valid NSF grant number.
• Curate it all the way through into the archive. Verify that everything still behaves as expected.
• Go to the item's page. Note the DOI. You should see the updated package page with the new information at the top and the funding info at the bottom, with the metadata.
• Verify that the Datacite metadata for the item contains the funding info:

```
xsltproc dryad-repo/dspace/config/crosswalks/DIM2DATACITE.xsl http://localhost:9999/resource/doi:<your_DOI>/mets.xml > test_datacite.xml
```

• Validate the metadata against the schema:

```
xmllint --noout --schema http://schema.datacite.org/meta/kernel-2.2/metadata.xsd test_datacite.xml
```

Probably it would be good to also test a sponsored-journal submission and a non-sponsored submission and make sure those still act the appropriate way.

daisieh commented:
Oh! Don't forget that you need to set the new NSF funder concept to have a `organization.paymentPlanType` of `DEFERRED`.

daisieh committed:
Merge remote-tracking branch 'datadryad/dryad-master' into dryad-master

ryscher commented:
Tested on VM.

Concepts are created correctly. Credit is properly applied for packages associated with NSF grants.

ryscher commented:
On the item page, package metadata is not displayed. The metadata for files and the citation are correct, but there is an issue generating the display of package's metadata.
<img width="751" alt="screenshot 2016-11-17 00 43 57" src="https://cloud.githubusercontent.com/assets/605143/20377668/22c32e98-ac5f-11e6-8b62-e326012c8ad2.png">

ryscher commented:
On the "show full metadata" screen, the funder information is displayed correctly, but is tagged with an undesirable warning icon.
<img width="705" alt="screenshot 2016-11-17 00 40 59" src="https://cloud.githubusercontent.com/assets/605143/20377699/4d97947e-ac5f-11e6-8fc8-0fa1d8f94f14.png">

ryscher commented on dspace/modules/payment-system/payment-api/src/main/java/org/dspace/workflow/actions/processingaction/FinalPaymentAction.java:
```diff
@@ -79,39 +80,35 @@ public ActionResult execute(Context c, WorkflowItem wfi, Step step, HttpServletR
+            DryadOrganizationConcept organizationConcept = shoppingCart.getSponsoringOrganization(c);
+            // if journal-based subscription is in place, transaction is paid
+            if (organizationConcept != null) {
+                if (shoppingCart.getJournalSub()) {
```
This line is a little hard to read. Can we change the name of getJournalSub() to hasSubscription() ?

daisieh replied:
Changed now.

ryscher commented:
Everything else looks good.

daisieh commented:
AUGH one problem that I am noting is that this PR is now woefully behind, Spring-wise, so I need to merge those changes in and then we can re-test.

daisieh committed:
Merge remote-tracking branch 'origin/spring-update' into temporary-branch

daisieh committed:
fix conflicts between this branch and dryad-master

daisieh commented:
Some problems with your testing: 
1) when you were looking at the package page, were you looking at the "internal-item" view or the "resource/doi" view? The funding information doesn't show up on the internal-item view, but it does show on the public view, so I think that should be sufficient. See first image below.

2) You need to use a valid grant number in order to get a non-scary icon. See second image below.

![screen shot 2016-11-18 at 4 46 13 pm](https://cloud.githubusercontent.com/assets/1827728/20451470/9e6f5a9e-adae-11e6-9f66-79af2b89923a.png)

![screen shot 2016-11-18 at 4 48 35 pm](https://cloud.githubusercontent.com/assets/1827728/20451497/e02b5c26-adae-11e6-8519-68bd23daf783.png)

daisieh committed:
hasSubscription is more accurate

made getJournalSub a private method, made a public hasSubscription
method.

daisieh committed:
Add funding information to internal item view

daisieh commented:
I've added the funding information to the internal item view as well.

daisieh committed:
rename methods for clarity

daisieh committed:
set the sponsoringOrganization for older carts that haven't had one set yet

ryscher commented:
I think my issues might be arising from the way I resolved the merge conflicts. For my current merge attempt, see #1250 -- this merge doesn't build, but I haven't looked into the issue yet.

daisieh committed:
clean merged versions of conflict files

# Conflicts:
#
dspace/modules/payment-system/payment-api/src/main/java/org/dspace/workf
low/actions/processingaction/FinalPaymentAction.java
#
dspace/modules/xmlui/src/main/java/org/datadryad/dspace/xmlui/aspect/bro
wse/ItemViewer.java

daisieh committed:
resolve conflicts

ryscher commented:
Tested updates on VM. Everything looks good. This is ready to merge any time.
